#!/usr/bin/env Rscript
library(optparse)

option_list = list(
  make_option(c("-f", "--file"), type="character", default=NULL, 
              help="assembly_metaquast_ranking.tsv file generated by QuasiModo", metavar="character"),
	make_option(c("-o", "--out"), type="character", default="assembly_metaquast_radarplot.pdf", 
              help="output file name of the radar-plot [default= %default]", metavar="character")
); 
 
opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);

if (is.null(opt$file)){
  print_help(opt_parser)
  stop("At least one argument must be supplied (input file)", call.=FALSE)
}

library(ggradar)
library(dplyr)
library(tidyr)
library(ggplot2)

assembler_metrics <- read.table(opt$file, sep="\t", header=T)
big_better <- c("Genome_fraction", "Largest_alignment", "NGA50")
small_better <- c("Duplication_ratio", "num_contigs", "num_mismatches_per_100_kbp")


assembler_metrics %>% 
  filter(criteria %in% big_better) %>% 
  group_by(criteria) %>% 
  mutate(min = min(average, na.rm=T), max = (max(average, na.rm=T))) %>%
  mutate(radar = (average - min)/(max-min)) ->
  assembly_big_better

assembler_metrics %>% 
  filter(criteria %in% small_better) %>% 
  group_by(criteria) %>% 
  mutate(min = min(average, na.rm=T), max = (max(average, na.rm=T))) %>%
  mutate(radar = (max - average)/(max-min)) ->
  assembly_small_better

rbind(assembly_big_better, assembly_small_better) %>% 
  select(assembler, criteria, radar) %>%
  spread(criteria, radar) %>%
  rename("Genome fraction" = "Genome_fraction", 
         "Largest alignment" = "Largest_alignment",
         "Duplication ratio" = "Duplication_ratio",
         "number contigs" = "num_contigs",
         "mismatches"="num_mismatches_per_100_kbp") ->
  assembly_radar 

radar_av_metric <- ggradar(assembly_radar)

ggsave(opt$out, width=10, height=7)
